### ISSUE TITLE
Allow matching of specific reference resolvement depth

### ISSUE THREAD

**Issue by jonaslagoni** [stoplightio/spectral#2120]
**User story.**
As a user, I want to decide the level of reference matching for a rule.

**Is your feature request related to a problem?**
In AsyncAPI we can define messages as references:
```yml
---
channels:
  some/channel:
    subscribe:
      message:
        $ref: ./some/message.yml
```

And the message object that is being referenced, can also contain a reference.

```yml
---
payload:
  $ref: ./some/schema.yml
```

I am unable to lint the reference within the referenced message object because I am not able to control the depth at which references should/should not be resolved. 

**Describe the solution you'd like**
The only way I can think this can be solved is through a depth level that you can "target" reference resolvement 🤔 Not quite sure how easy it is to achieve though!

For example, given:
```yml
rules:
  ...
  asyncapi-message-payload-must-use-references:
    given: $.channels.[*][subscribe,publish].message
    resolved: "depth-1"
```

It would resolve the first reference only:
```yml
---
channels:
  some/channel:
    subscribe:
      message:
        payload:
          $ref: ./some/schema.yml
```

Maybe someone have a better way to achieve this? 🤔

**Comment by SirAppSec** [comment]
I'm facing a similar requirement for some specific internal rules, @jonaslagoni, did you find a solution?
I considered custom resolvers, but they seem to require file resolving and this can be a pain to achieve properly.

In my case I just need to know if an object is a de-refrenced object and the ref itself.
I manage to achieve this for the first depth using context.document but it only show the first "depth".
Is it possible to get a graph of references?

**Comment by jonaslagoni** [comment]
@SirAppSec I have unfortunately not no.
